File: test/ModularPipelines.UnitTests/Attributes/ModuleReadyEventTests.cs
using ModularPipelines.Attributes;
using ModularPipelines.Attributes.Events;
using ModularPipelines.Context;
using ModularPipelines.Enums;
using ModularPipelines.Modules;
using ModularPipelines.Modules.Behaviors;
using ModularPipelines.TestHelpers;

namespace ModularPipelines.UnitTests.Attributes;

[TUnit.Core.NotInParallel(nameof(ModuleReadyEventTests))]
public class ModuleReadyEventTests : TestBase
{
    private static readonly List<string> EventLog = new();

    public class LogReadyAttribute : Attribute, IModuleReadyEventReceiver
    {
        public Task OnModuleReadyAsync(IModuleReadyContext context)
        {
            EventLog.Add($"Ready:{context.ModuleName}");
            return Task.CompletedTask;
        }
    }

    public class LogReadyWithTimingAttribute : Attribute, IModuleReadyEventReceiver
    {
        public Task OnModuleReadyAsync(IModuleReadyContext context)
        {
            EventLog.Add($"Ready:{context.ModuleName}:WaitTime:{context.TimeWaitingForDependencies.TotalMilliseconds >= 0}");
            return Task.CompletedTask;
        }
    }

    public class LogReadyAndStartAttribute : Attribute, IModuleReadyEventReceiver, IModuleStartEventReceiver
    {
        public Task OnModuleReadyAsync(IModuleReadyContext context)
        {
            EventLog.Add($"Ready:{context.ModuleName}");
            return Task.CompletedTask;
        }

        public Task OnModuleStartAsync(IModuleEventContext context)
        {
            EventLog.Add($"Start:{context.ModuleName}");
            return Task.CompletedTask;
        }
    }

    public class ThrowingReadyAttribute : Attribute, IModuleReadyEventReceiver
    {
        public bool ContinueOnError => true;

        public Task OnModuleReadyAsync(IModuleReadyContext context)
        {
            throw new InvalidOperationException("Ready event failed");
        }
    }

    [LogReady]
    public class SimpleModuleWithReadyEvent : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            return Task.FromResult<string?>("Done");
        }
    }

    [LogReadyWithTiming]
    public class ModuleWithTimingCheck : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            return Task.FromResult<string?>("Done");
        }
    }

    [LogReadyAndStart]
    public class ModuleWithReadyAndStart : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            return Task.FromResult<string?>("Done");
        }
    }

    [LogReady]
    [ModularPipelines.Attributes.DependsOn<DependencyModule>]
    public class DependentModuleWithReadyEvent : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            return Task.FromResult<string?>("Dependent Done");
        }
    }

    public class DependencyModule : Module<string>
    {
        public override async Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            await Task.Delay(10, cancellationToken);
            return "Dependency Done";
        }
    }

    [ThrowingReady]
    public class ModuleWithThrowingReadyEvent : Module<string>
    {
        public override Task<string?> ExecuteAsync(IModuleContext context, CancellationToken cancellationToken)
        {
            return Task.FromResult<string?>("Done");
        }
    }

    [Before(Test)]
    public void ClearEventLog()
    {
        EventLog.Clear();
    }

    [Test]
    public async Task ReadyEvent_IsFired_WhenModuleIsReady()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<SimpleModuleWithReadyEvent>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);
        await Assert.That(EventLog).Contains("Ready:SimpleModuleWithReadyEvent");
    }

    [Test]
    public async Task ReadyEvent_ProvidesValidContext()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<ModuleWithTimingCheck>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);
        await Assert.That(EventLog.Any(e => e.Contains("Ready:ModuleWithTimingCheck:WaitTime:True"))).IsTrue();
    }

    [Test]
    public async Task ReadyEvent_FiresBeforeStartEvent()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<ModuleWithReadyAndStart>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);

        var readyIndex = EventLog.IndexOf("Ready:ModuleWithReadyAndStart");
        var startIndex = EventLog.IndexOf("Start:ModuleWithReadyAndStart");

        await Assert.That(readyIndex).IsGreaterThanOrEqualTo(0);
        await Assert.That(startIndex).IsGreaterThanOrEqualTo(0);
        await Assert.That(readyIndex).IsLessThan(startIndex);
    }

    [Test]
    public async Task ReadyEvent_IsFired_WhenDependenciesComplete()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<DependencyModule>()
            .AddModule<DependentModuleWithReadyEvent>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);
        await Assert.That(EventLog).Contains("Ready:DependentModuleWithReadyEvent");
    }

    [Test]
    public async Task ReadyEvent_WithContinueOnError_DoesNotFailPipeline()
    {
        var result = await TestPipelineHostBuilder.Create()
            .AddModule<ModuleWithThrowingReadyEvent>()
            .ExecutePipelineAsync();

        await Assert.That(result.Status).IsEqualTo(Status.Successful);
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Attributes/ModuleReadyEventTests.fs
namespace ModularPipelines.UnitTests.FSharp.Attributes

open System
open System.Threading
open System.Threading.Tasks
open ModularPipelines.Attributes
open ModularPipelines.Attributes.Events
open ModularPipelines.Context
open ModularPipelines.Enums
open ModularPipelines.Extensions
open ModularPipelines.Modules
open ModularPipelines.TestHelpers
open TUnit.Assertions
open TUnit.Assertions.Extensions
open TUnit.Assertions.FSharp.Operations
open TUnit.Core

module ModuleReadyEventTests =
    let eventLog = ResizeArray<string>()

    type LogReadyAttribute() =
        inherit Attribute()
        interface IModuleReadyHandler with
            member _.ContinueOnError = false
            member _.OnModuleReadyAsync(context: IModuleHookContext) =
                eventLog.Add($"Ready:{context.ModuleName}")
                Task.CompletedTask

    type LogReadyWithTimingAttribute() =
        inherit Attribute()
        interface IModuleReadyHandler with
            member _.ContinueOnError = false
            member _.OnModuleReadyAsync(context: IModuleHookContext) =
                eventLog.Add($"Ready:{context.ModuleName}:ElapsedTime:{context.ElapsedTime.TotalMilliseconds >= 0.0}")
                Task.CompletedTask

    type LogReadyAndStartAttribute() =
        inherit Attribute()
        interface IModuleReadyHandler with
            member _.ContinueOnError = false
            member _.OnModuleReadyAsync(context: IModuleHookContext) =
                eventLog.Add($"Ready:{context.ModuleName}")
                Task.CompletedTask
        interface IModuleStartHandler with
            member _.ContinueOnError = false
            member _.OnModuleStartAsync(context: IModuleHookContext) =
                eventLog.Add($"Start:{context.ModuleName}")
                Task.CompletedTask

    type ThrowingReadyAttribute() =
        inherit Attribute()
        interface IModuleReadyHandler with
            member _.ContinueOnError = true
            member _.OnModuleReadyAsync(_: IModuleHookContext) =
                raise (InvalidOperationException("Ready event failed"))

    [<LogReady>]
    type SimpleModuleWithReadyEvent() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("Done")

    [<LogReadyWithTiming>]
    type ModuleWithTimingCheck() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("Done")

    [<LogReadyAndStart>]
    type ModuleWithReadyAndStart() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("Done")

    type DependencyModule() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, ct: CancellationToken) =
            task {
                do! Task.Delay(10, ct)
                return "Dependency Done"
            }

    [<LogReady>]
    [<DependsOn(typeof<DependencyModule>)>]
    type DependentModuleWithReadyEvent() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("Dependent Done")

    [<ThrowingReady>]
    type ModuleWithThrowingReadyEvent() =
        inherit Module<string>()
        override _.ExecuteAsync(_: IModuleContext, _: CancellationToken) =
            Task.FromResult<string>("Done")

[<TUnit.Core.NotInParallel(nameof ModuleReadyEventTests)>]
type ModuleReadyEventTests() =
    inherit TestBase()

    [<Before(HookType.Test)>]
    member _.ClearEventLog() =
        ModuleReadyEventTests.eventLog.Clear()

    [<Test>]
    member _.ReadyEvent_IsFired_WhenModuleIsReady() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<ModuleReadyEventTests.SimpleModuleWithReadyEvent>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status = Status.Successful).IsTrue())
        do! check(Assert.That(ModuleReadyEventTests.eventLog |> Seq.contains "Ready:SimpleModuleWithReadyEvent").IsTrue())
    }

    [<Test>]
    member _.ReadyEvent_ProvidesValidContext() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<ModuleReadyEventTests.ModuleWithTimingCheck>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status = Status.Successful).IsTrue())
        do! check(Assert.That(ModuleReadyEventTests.eventLog |> Seq.exists (fun e -> e.Contains("Ready:ModuleWithTimingCheck:ElapsedTime:True"))).IsTrue())
    }

    [<Test>]
    member _.ReadyEvent_FiresBeforeStartEvent() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<ModuleReadyEventTests.ModuleWithReadyAndStart>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status = Status.Successful).IsTrue())

        let readyIndex = ModuleReadyEventTests.eventLog.IndexOf("Ready:ModuleWithReadyAndStart")
        let startIndex = ModuleReadyEventTests.eventLog.IndexOf("Start:ModuleWithReadyAndStart")

        do! check(Assert.That(readyIndex).IsGreaterThanOrEqualTo(0))
        do! check(Assert.That(startIndex).IsGreaterThanOrEqualTo(0))
        do! check(Assert.That(readyIndex).IsLessThan(startIndex))
    }

    [<Test>]
    member _.ReadyEvent_IsFired_WhenDependenciesComplete() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<ModuleReadyEventTests.DependencyModule>()
                .AddModule<ModuleReadyEventTests.DependentModuleWithReadyEvent>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status = Status.Successful).IsTrue())
        do! check(Assert.That(ModuleReadyEventTests.eventLog |> Seq.contains "Ready:DependentModuleWithReadyEvent").IsTrue())
    }

    [<Test>]
    member _.ReadyEvent_WithContinueOnError_DoesNotFailPipeline() = async {
        let! result =
            TestPipelineHostBuilder.Create()
                .AddModule<ModuleReadyEventTests.ModuleWithThrowingReadyEvent>()
                .ExecutePipelineAsync()
            |> Async.AwaitTask

        do! check(Assert.That(result.Status).IsEqualTo(Status.Successful))
    }
